WIP: Metrics: Add server for Egress IP/firewall

martinkennelly opened the pull request:
Add an insecure HTTP metrics server that serves
metrics on loopback port 29100.

Add metrics to count egress firewalls,
egress firewall rules and egress IPs count.

In order to validate the metrics:

Metric `num_egress_ips`:
1. Configure EgressIP: https://docs.openshift.com/container-platform/4.9/networking/openshift_sdn/assigning-egress-ips.html
2. Exec into SDN pod that is the current LE leader. You can see whos the leader in a config map in the SDN namespace.
3. curl localhost:29100/metrics | grep num_egress_ips
4. You should see an integer from step 3 which represents the total number of egress IPs. This should be the same as the amount you created in step 1.

Metric `num_egress_firewall_rules`:
1. Configure EgressFirewall: https://docs.openshift.com/container-platform/4.9/networking/openshift_sdn/configuring-egress-firewall.html
2. Exec into SDN pod that is the current LE leader. You can see whos the leader in a config map in the SDN namespace.
3. curl localhost:29100/metrics | grep num_egress_firewall_rules
4. You should see an integer from step 3 which represents the total number of EgressFirewall egress rules. This should be the same as the amount you created in step 1.

Metric `num_egress_firewalls`: 
1. 1. Configure X number EgressFirewall: https://docs.openshift.com/container-platform/4.9/networking/openshift_sdn/configuring-egress-firewall.html
2. Exec into SDN pod that is the current LE leader. You can see whos the leader in a config map in the SDN namespace.
3. curl localhost:29100/metrics | grep num_egress_firewalls
4. You should see an integer from step 3 which represents the total number of EgressFirewalls. This should be the same as the amount you created in step 1.


Signed-off-by: Martin Kennelly <mkennell@redhat.com>

danwinship commented on pkg/network/master/egressip.go:
```diff
@@ -240,3 +243,90 @@ func (eim *egressIPManager) SetNamespaceEgressDropped(vnid uint32) {
+
+type egressNetworkPolicyCache struct {
+	*sync.Mutex
+	// namespace name -> egress network policy name -> egressNetworkPolicy
```
FTR, it's illegal to have more than one EgressNetworkPolicy in a namespace, so you could remove a level here

martinkennelly commented on pkg/network/master/egressip.go:
```diff
@@ -240,3 +243,90 @@ func (eim *egressIPManager) SetNamespaceEgressDropped(vnid uint32) {
 }
+
+type egressNetworkPolicyCache struct {
+	*sync.Mutex
```
I put it as a pointer so anyone can pass this struct by value or reference to a function and no issue with copy lock. Why is it normally non-pointers?

martinkennelly commented:
/retest

martinkennelly commented:
/retest
CI flaking. Submitted BZ for flake: https://bugzilla.redhat.com/show_bug.cgi?id=2019001

danwinship commented on pkg/network/master/metrics/register.go:
```diff
@@ -0,0 +1,32 @@
+	Namespace: metricSDNNamespace,
+	Subsystem: metricSDNSubsystemController,
+	Name:      "num_egress_firewall_rules",
+	Help:      "The number of egress firewall rules defined"},
```
Can you add a metric for number of distinct ENPs too? I think both would be useful.

danwinship commented on pkg/network/master/egressnp.go:
```diff
@@ -0,0 +1,47 @@
+package master
```
can you rename this file to `egress_network_policy.go` so it matches the one in `pkg/network/node`?

danwinship commented on pkg/network/master/egressnp.go:
```diff
@@ -0,0 +1,47 @@
+func (enp *egressNetworkPolicyManager) handleAddUpdate(current, old interface{}, _ watch.EventType) {
+	var change int
+	currentEgressNetworkPolicy, _ := current.(*osdnv1.EgressNetworkPolicy)
+	if old != nil {
+		oldEgressNetworkPolicy, _ := old.(*osdnv1.EgressNetworkPolicy)
+		change = len(currentEgressNetworkPolicy.Spec.Egress) - len(oldEgressNetworkPolicy.Spec.Egress)
+	} else {
+		change = len(currentEgressNetworkPolicy.Spec.Egress)
+	}
+	enp.egressRuleCount += change
```
You know what would go great with that code? A unit test! :slightly_smiling_face:

danwinship reviewed (approved):
/approve
general idea is good, needs a few tweaks, but then anyone can /lgtm it after that

martinkennelly commented:
@danwinship I made the changes you requested - ENP metric, renamed file, and unit tests.
During my testing, I decided to change the definition of `num_egress_ips` to count egress IPs assigned to nodes [1]. This is better than counting all because it lets us know how many egress IPs are currently *active* on a cluster. I updated the HELP field to inform consumers of this distinction.

I also noticed an error I made in the original code you reviewed. I made the mistake of checking if interface{} was nil for func parameter `old` in `handleAddUpdate`. I now check the event type to find out if its an add or update operation [2].

[1] https://github.com/openshift/sdn/pull/358/files#diff-cb8e1be9489f2925e33943b9051dded85e22f36cb5b9905ebeb807f820cb9627R700

[2] https://github.com/openshift/sdn/pull/358/files#diff-2153a20dc3cd4885ce75eac5c9e4a51f6343d6be8b0d154be9973c1e66bad0a0R30

martinkennelly commented:
/retest
CI flakes. Added [BZ ](https://bugzilla.redhat.com/show_bug.cgi?id=2020195)for one of them.

alexanderConstantinescu reviewed (commented):
This is really an lgtm, I just had a small nit that I wanted to give you time to decide what to do about (and if I put a real lgtm, the bot will merge the PR). Let me know

martinkennelly commented on pkg/network/master/metrics/register.go:
```diff
@@ -0,0 +1,41 @@
+var metricEgressIPCount = prometheus.NewGauge(prometheus.GaugeOpts{
+	Namespace: metricSDNNamespace,
+	Subsystem: metricSDNSubsystemController,
+	Name:      "num_egress_ips",
```
Id love to change it to what you suggested but I want to keep this name consistent with OVN-K and that PR is merged. Its not worth it for me to rename it now.

alexanderConstantinescu replied:
Understood then :)
